### DISCUSSION TITLE
Consolidate haunted house and crooked house operating modes

### DISCUSSION THREAD

**Discussion by spacek531** [OpenRCT2/OpenRCT2#18967] [Feature requests]
A continuation of https://github.com/OpenRCT2/OpenRCT2/discussions/18966

Haunted houses have some interesting hardcoded behavior that could be un-hardcoded, which would allow crooked house and haunted house modes to be functionally identical.

Here is its update code:
https://github.com/OpenRCT2/OpenRCT2/blob/279675ba45f4607b211a074f3a331cc4fe0f2551/src/openrct2/ride/Vehicle.cpp#L4961-L5009

First of all, the haunted house has only a single animation lasting 19 frames (frames 1 - 19), which is activated twice in the ride cycle, with each frame playing twice. The rest of the haunted house mode is playing sound effects at various times.

A combination of animation maps and a sequence of events could allow the haunted house to be customized much further. Consider the following:

```cpp
struct HauntedHouseEvent = {
    uint16_t eventTime = 0;
    uint8_t targetSubState = 0;
    OpenRCT2::Audio::SoundId soundEffect = OpenRCT2::Audio::SoundId::Null;
};
```

This struct would be used to trigger animations and sound effects at specific times in the ride cycle. `targetSubState` would force the vehicle's sub_state to the desired sub_state, which would cause an animation to play using the ubiquitous time-to-sprite map system. When the animation is finished playing, it would revert to sub_state 0 automatically and wait for the next event. `soundEffect` obviously plays a sound effect. Sub state 0 would be a looping animation. A value of 255 in `targetSubState` could indicate the ride ending, or a separate bool value could be used in each struct.

In the case of the vanilla haunted house, it would have a map and sequence something like this:
(only using 1-d maps for simplicity)
```cpp
using SpriteMapValue = uint8_t;

std::vector<std::vector<SpriteMapValue>> HauntedHouseSubstates = {
    { 0 },
    { 1, 1, 2, 2, 3, 3, 4, 4, 5, 5, 6, 6, 7, 7, 8, 8, 9, 9, 10, 10, 11, 11, 12, 12, 13, 13, 14, 14, 15, 15, 16, 16, 17, 17, 18, 18, 19, 19, 255},
 };

std::vector<HauntedHouseEvent> HauntedHouseEvents = {
    { 44, 0, 45 }, { 74, 1, 255 }, { 399, 0, 46 }, { 744, 0, 45 }, { 754, 1, 255 }, { 1199, 0, 47 }, { 1500, 255, 255}
};
 ```
 These would be used something like this:
 ```cpp
void Vehicle::UpdateHauntedHouseOperating()
{
    if (_vehicleBreakdown == 0)
        return;
    
    // Determine the most recent event
    int8_t currentEventNumber = -1;
    HauntedHouseEvent currentEvent{}; // null event
    for (auto &event const = HauntedHouseEvents.rbegin(); event != HauntedHouseEvents.rend(); ++event)
    {
        if (Event.eventTime <= current_time)
        {
            currentEvent = event;
            break;
         }
    }
    // If it is the first frame of the event, change substate and play a sound if applicable
    if (currentEvent.eventTime - current_time == 0)
    {
        sub_state = currentEvent.targetSubState
        if ( currentEvent.soundEffect != OpenRCT2::Audio::SoundId::Null )
            OpenRCT2::Audio::Play3D(currentEvent.soundEffect, GetLocation());
    }
    
    // If the event says to quit, quit.
    if (currentEvent.targetSubState == 255)
    {
        SetState(Vehicle::Status::Arriving);
        var_C0 = 0;
        return;
    }
    
    // Choose the animation frame to play
    auto subStateTime = current_time - currentEvent.eventTime;
    auto animationMap = HauntedHouseSubstates[sub_state];
    // animations without a defined end loop until substate is forced to change by an event
    subStateTime %= animationMap.size;
    auto animationFrame = animationMap[subStateTime];
    
    // If the animation has ended, switch to the first frame of the base substate
    if (animationFrame == 255)
    {
        sub_state = 0;
        animationFrame = HauntedHouseSubstates[sub_state][0];
    }
    
    // If the animation is playing, invalidate
    if (animationFrame != pitch)
    {
        pitch = animationFrame;
        Invalidate();
    }
    
    current_time++;
}
 ```
 (this is completely untested off the seat of my pants. It ignores some of the previously proposed changes like the 2D sprite map)

The eventTime values are 1 below the vanilla counterparts because current_time is incremented before the switch statement i.e. it is one-index while my implementation increments after the event i.e. zero-index.  The value for ride cycle end (1500) is not affected because the check is before increment.

There is a potential desync related to gCurrentTick. The haunted house only increments frames on the odd ticks, so if current_time == 74 on an odd tick the first animation frame will last 2 ticks but if current_time == 74 on an even tick the animation frame will last only 1 tick. If the game imports an old save with the a haunted house whose even current_time occur on an even gCurrentTick, the haunted house animation will play 1 tick behind a vanilla counterpart until the animation is finished, which I believe is of no consequence.

 I think circus could also be rolled into this operating mode. If exposed to ride objects, this event system would be extremely powerful and allow for a lot of unique "show" rides.

Consider for example a dolphin show. A ride creator could make an entire show out of smaller animation segments. The default animation could be a simple 8-frame animation of the dolphin circling the tank. There would be animations for the dolphin doing different jumps or tail walking. The show sequence would then be created by adding events at the appropriate times to make a seamless animation.

**Comment by ocalhoun6** [comment]
The 3D Cinema could probably also be consolidated into this same operating mode.

Possibly even the Motion Simulator as well?

**Reply by spacek531** [reply]
Motion simulator easily yes since there is only one animation. 3D cinema would be more difficult since there are different film durations based on operating mode - it would require show sequences for each operating mode.

Space rings could also be consolidated.

**Reply by ocalhoun6** [reply]
Motion Simulator also has multiple ride durations based on which film you choose. So maybe not, then.

But I'd think that multiple show types/durations would be something you'd want to incorporate into the consolodated 'show' ride type anyway. Like, if you were going to make a new object based on that ride type for a fireworks or laser light show, you'd probably also want to be able to choose between different show modes of different durations, maybe even with different animations. Or a 'Stage Show' object might have a few different options of what kind of show to put on the stage, etc.

**Reply by spacek531** [reply]
I know it has two operating modes but the source code suggests they act 100% identical wrt ride duration
https://github.com/OpenRCT2/OpenRCT2/blob/279675ba45f4607b211a074f3a331cc4fe0f2551/src/openrct2/ride/Vehicle.cpp#L4814
The ride uses a single map, and the map determines the ride ending (the 255 check).

**Comment by spacek531** [comment]
Thinking about it, things can be consolidated more. Any flat ride that has a fixed ride sequence per operating mode can be consolidated into the events system:
- Space Rings
- 3D Cinema
- Motion Simulator
- Top Spin
- Haunted House
- Crooked House
- maybe Spiral Slide

For these rides it would be as simple as an event starting the sequence then an event telling the ride to stop:
```cpp
// motion simulator
std::vector<HauntedHouseEvent> AvengingAviatorsEvents = {{0, 0, 255}, {1643, 255, 255}};
std::vector<HauntedHouseEvent> ThrillRidersEvents = {{0, 0, 255}, {1643, 255, 255}};
// top spin
std::vector<HauntedHouseEvent> BeginnerEvents= {{0, 0, 255}, {562, 255, 255}};
std::vector<HauntedHouseEvent> IntermediateEvents= {{0, 1, 255}, {698, 255, 255}};
std::vector<HauntedHouseEvent> BerserkEvents= {{0, 2, 255}, {1643, 944, 255}};
// 3D cinema
std::vector<HauntedHouseEvent> MouseTailsEvents= {{0, 0, 255}, {5000, 255, 255}};
std::vector<HauntedHouseEvent> StormChasersEvents = {{0, 0, 255}, {6000, 255, 255}};
std::vector<HauntedHouseEvent> SpaceRaidersEvents= {{0, 0, 255}, {7000, 255, 255}};
// Space Rings
std::vector<HauntedHouseEvent> SpaceRingsEvents = {{0, 0, 255}, {2062, 255, 255}};)
```
